_posts/2021-07-13-JEV.md

```diff
@@ -0,0 +1,822 @@
+        - [Monkey Wrench Drop - where AV fails and EV thrives](#monkey-wrench-drop---where-av-fails-and-ev-thrives)
+      - [Real world example](#real-world-example)
+        - [Problem statement](#problem-statement)
+        - [Vanilla MinMax() version](#vanilla-minmax-version)
+        - [SIMD MinMax() version](#simd-minmax-version)
+          - [Elephant in the room](#elephant-in-the-room)
+  - [Epilogue](#epilogue)
+- [References](#references)
+
+## Part 1 - background (written by dummies, for curious cats)
+
+### What is _'Vectorization'_?
+
+Without beating about the bush, does Vectorization, explicit or otherwise, have anything to do with the
+fantabulous `java.util.Vector` class? Well, not much.
+
+Allow me to explain with this very long example.
+
+#### Bullshit use-case
+
+Let's say you are writing an image editing tool similar to Photoshop. One of the functions you want to implement is
+adjusting the brightness. In this particular example:
+
+- Your input is an array of 3 bytes of `RGB` values per pixel, meaning a `1920 x 1080` picture would
+  have `3 x 1920 x 1080 = 6220800` bytes.
+- Increasing brightness is basically adding 1 (or more) to each RGB value per pixel
+
+So how would we do that?
+
+```java
+  for (int i = 0; i < 6220800; i++) {
+      rgb[i]++;
+  }
+```
+
+We could probably do something fancy like unrolling the loop (eliminating some bounds checking) to speed this up:
+
+```java
+  for (int i = 0; i < 6220800; i += 2) {
+      rgb[i + 0]++;
+      rgb[i + 1]++;
+  }
+```
+
+The reason we have to do it this way, is because of real world physics. Your fancy shmancy CPU is basically
+a `digital logic gate` on steroids, that can do A LOT of ~~dumb~~ simple things really REALLY fast. 
+
+##### Vivisection
+
+To give you an idea of what "simple" means, if you want to add 2 numbers you have to:
+
+- 1st simple operation (`mov reg1, num1`): load first number into register (small memory inside CPU)
+- 2nd simple operation (`add reg1, num2`): add second number to register
+- 3rd simple operation (`mov num1, reg1`): write register back to first number
+
+That is 3 CPU operations for something that seems rather innocent. The good news is, is that your CPU (depending on
+speed, and the type of ops) can do millions of these operations per second.
+
+Let's say our CPU can do a million operations per second. Let us do a simplified dissection of the first example from
+above:
+
+- 3 instructions: save current state
+- 3 instructions: setting up the loop
+    - 3 instructions: checking/incrementing the loop
+    - 3 instructions: adding 1 to a value (`mov, add, mov`)
+- 3 instructions: restore saved state
+
+So about 6 instructions per loop iteration. Multiply that by the size of our loop `6 x 6220800 = 37324800`, which when
+divided by a million operations per second results in `37324800 / 1000000 = 37` seconds to increase the brightness of
+a `1920 x 1080` picture (on a fundamental level, computers have a mechanical nature to them).
```

Comment:
on a fundamental level, computers have a mechanical nature to them -> what do you mean with this? That it doesn' necessarily translate into a (1 linear step) increase in brightness?